### ISSUE TITLE
Identity publish failure

### ISSUE THREAD

**Issue by franckc** [OriginProtocol/origin#2623]
I observed this issue while testing with the Coinbase mobile wallet but @tomlinton also observed it on desktop so this does not seem to be wallet or device specific.

The steps were as follow:
 - Update my profile
 - Press publish
 - Signed transaction with data sent to relayer
 - Modal for waiting on blockchain with spinner showed up
 - After about 30 second, got error "Error - see console".
My wallet address is: 0xc5c06bb14161c98a21b178bE9ACb77169eCc5C9D.

I checked the logs in the relayer and did not see anything suspicious. The tx was mined successfully. The relayer received the transaction at 2019-07-04 02:44:29 UTC

When I reloaded the page, the data that I was publishing showed up. So the data was published and persisted in blockchain/IPFS correctly.

Here is the screenshot from Tom that he captured on desktop:

<img width="1344" alt="Screen_Shot_2019-07-04_at_2 55 40_PM" src="https://user-images.githubusercontent.com/945910/60638648-0ab87880-9dd4-11e9-820f-8838390c2b70.png">

**Comment by franckc** [comment]
CC @nick @shahthepro @mikeshultz In case they have any idea what the issue could be...

**Comment by franckc** [comment]
Happened again for me. Interestingly, this time the "Error - see console modal" showed up first after I signed the transaction, then got replaced by the "Waiting on blockchain modal" and then it showed my profile as updated successfully...

**Comment by mikeshultz** [comment]
Interestingly, here's the result of the transaction you sent at 4:15UTC:

    {
      "data": {
        "web3": {
          "blockNumber": 8082775,
          "transactionReceipt": {
            "id": "0x3ebfa40f4c92a239c0ee8bd149fb8a8e2d36a7908a678033a065c105013a2f70",
            "blockNumber": 8082692,
            "events": [
              {
                "id": "log_ec277ff9",
                "event": null,
                "returnValues": null,
                "returnValuesArr": null
              },
              {
                "id": "log_e692bd43",
                "event": "IdentityUpdated",
                "returnValues": null,
                "returnValuesArr": [
                  {
                    "field": "account",
                    "value": "0x7CEf86cED36DdA57dc6D7253eA540f941eFbB3fD"
                  },
                  {
                    "field": "ipfsHash",
                    "value": "0xb5444e2c35169cabf2d6cdc057b65a6846a8749cef1da9f4fb82a66049d0f3e8"
                  }
                ]
              },
              {
                "id": "log_4aa784ab",
                "event": null,
                "returnValues": null,
                "returnValuesArr": null
              }
            ]
          }
        }
      }
    }

That last event is... special.

**Comment by mikeshultz** [comment]
If someone that triggers this could grab any/all of the JSON-RPC network calls made(I think they should be local), that might be helpful.  There's a chance bad data from the provider could cause this in the resolver.

**Comment by mikeshultz** [comment]
There's a chance this isn't actually related and a bug in the postgres backend, but figure I'd add it.  While I look into this, it could shed some light onto the reported issue.

    event-listener    | [ERROR] event-listener: { Error: GraphQL error: Cannot return null for non-nullable field Event.id.
    event-listener    |     at new ApolloError (/app/node_modules/apollo-client/bundle.umd.js:92:26)
    event-listener    |     at /app/node_modules/apollo-client/bundle.umd.js:1581:34
    event-listener    |     at /app/node_modules/apollo-client/bundle.umd.js:2001:15
    event-listener    |     at Set.forEach (<anonymous>)
    event-listener    |     at /app/node_modules/apollo-client/bundle.umd.js:1999:26
    event-listener    |     at Map.forEach (<anonymous>)
    event-listener    |     at QueryManager.broadcastQueries (/app/node_modules/apollo-client/bundle.umd.js:1997:20)
    event-listener    |     at /app/node_modules/apollo-client/bundle.umd.js:2124:19
    event-listener    |     at Object.next (/app/node_modules/zen-observable/lib/Observable.js:308:23)
    event-listener    |     at notifySubscription (/app/node_modules/zen-observable/lib/Observable.js:130:18)
    event-listener    |     at onNotify (/app/node_modules/zen-observable/lib/Observable.js:165:3)
    event-listener    |     at SubscriptionObserver.next (/app/node_modules/zen-observable/lib/Observable.js:219:7)
    event-listener    |     at /app/node_modules/apollo-client/bundle.umd.js:1095:36
    event-listener    |     at Set.forEach (<anonymous>)
    event-listener    |     at Object.next (/app/node_modules/apollo-client/bundle.umd.js:1094:21)
    event-listener    |     at notifySubscription (/app/node_modules/zen-observable/lib/Observable.js:130:18)
    event-listener    |     at onNotify (/app/node_modules/zen-observable/lib/Observable.js:165:3)
    event-listener    |     at SubscriptionObserver.next (/app/node_modules/zen-observable/lib/Observable.js:219:7)
    event-listener    |     at /app/node_modules/apollo-link-schema/lib/index.js:24:30
    event-listener    |   graphQLErrors:
    event-listener    |    [ { Error: Cannot return null for non-nullable field Event.id.
    event-listener    |          at defaultMergedResolver (/app/packages/graphql/node_modules/graphql-tools/dist/stitching/defaultMergedResolver.js:16:36)
    event-listener    |          at field.resolve (/app/packages/graphql/node_modules/graphql-tools/dist/stitching/mergeSchemas.js:206:24)
    event-listener    |          at resolveFieldValueOrError (/app/node_modules/graphql/execution/execute.js:503:18)
    event-listener    |          at resolveField (/app/node_modules/graphql/execution/execute.js:470:16)
    event-listener    |          at executeFields (/app/node_modules/graphql/execution/execute.js:311:18)
    event-listener    |          at collectAndExecuteSubfields (/app/node_modules/graphql/execution/execute.js:747:10)
    event-listener    |          at completeObjectValue (/app/node_modules/graphql/execution/execute.js:737:10)
    event-listener    |          at completeValue (/app/node_modules/graphql/execution/execute.js:626:12)
    event-listener    |          at completeValueCatchingError (/app/node_modules/graphql/execution/execute.js:531:19)
    event-listener    |          at /app/node_modules/graphql/execution/execute.js:651:25
    event-listener    |          at Array.forEach (<anonymous>)
    event-listener    |          at forEach (/app/node_modules/iterall/index.js:83:25)
    event-listener    |          at completeListValue (/app/node_modules/graphql/execution/execute.js:647:24)
    event-listener    |          at completeValue (/app/node_modules/graphql/execution/execute.js:609:12)
    event-listener    |          at completeValueCatchingError (/app/node_modules/graphql/execution/execute.js:531:19)
    event-listener    |          at resolveField (/app/node_modules/graphql/execution/execute.js:471:10)
    event-listener    |          at executeFields (/app/node_modules/graphql/execution/execute.js:311:18)
    event-listener    |          at collectAndExecuteSubfields (/app/node_modules/graphql/execution/execute.js:747:10)
    event-listener    |          at completeObjectValue (/app/node_modules/graphql/execution/execute.js:737:10)
    event-listener    |          at completeAbstractValue (/app/node_modules/graphql/execution/execute.js:694:10)
    event-listener    |          at completeValue (/app/node_modules/graphql/execution/execute.js:621:12)
    event-listener    |          at completeValueCatchingError (/app/node_modules/graphql/execution/execute.js:531:19)
    event-listener    |          at resolveField (/app/node_modules/graphql/execution/execute.js:471:10)
    event-listener    |          at executeFields (/app/node_modules/graphql/execution/execute.js:311:18)
    event-listener    |          at collectAndExecuteSubfields (/app/node_modules/graphql/execution/execute.js:747:10)
    event-listener    |          at completeObjectValue (/app/node_modules/graphql/execution/execute.js:737:10)
    event-listener    |          at completeValue (/app/node_modules/graphql/execution/execute.js:626:12)
    event-listener    |          at /app/node_modules/graphql/execution/execute.js:528:16
    event-listener    |        message: 'Cannot return null for non-nullable field Event.id.',
    event-listener    |        locations: [],
    event-listener    |        path: [Array] } ],
    event-listener    |   networkError: null,
    event-listener    |   message:
    event-listener    |    'GraphQL error: Cannot return null for non-nullable field Event.id.',
    event-listener    |   extraInfo: undefined } 'will retry in 0.932 seconds, retry count 3'

**Comment by mikeshultz** [comment]
Yeah, the above error was due to the event ID not being stored, and in turn not returned.  @tomlinton since you saw this on the desktop, could you check your indexed db in local storage and see if the data looks sane(and includes the ID for all events)?

**Comment by franckc** [comment]
I'm seeing [this error](https://sentry.io/organizations/origin-protocol/issues/1096011500/?project=1399965&query=is%3Aunresolved&statsPeriod=14d) in Sentry which I think is related (though Sentry trims the error so I'm not 100% sure).

It is happening at a frequency of 2 to 10 events/day.

**Comment by mikeshultz** [comment]
That appears to be an exchange rate issue.

    Could not retrieve balances using GraphQL:  Error: TypeError: null is not an object (evaluating 'o.toString') Given value: "null"

**Comment by tomlinton** [comment]
That is a eth/token balance retrieval graphql query. That output in Sentry isn't a chain of events, it is more like a console log. Check the timestamps of the events. Those two issues aren't related to the core identity publish problem here and shouldn't have any negative impact.

The truncated error for the identity issue is:

`message: Cannot return null for non-nullable fiel…, `

Would be nice if I could update that to not truncate.

**Comment by franckc** [comment]
I suspect the issue might be in the getTransactionReceipt resolver function [here](https://github.com/OriginProtocol/origin/blob/master/packages/graphql/src/resolvers/web3/transactions.js#L6).
Somehow in rare cases the event may not have an id ? I have no idea why it would be the case... maybe I should just add a check in that method and if no id present for an event throw an exception. Any better idea ?

**Comment by mikeshultz** [comment]
I wonder if this is a different web3 coming into play.  Do we know if this only happens with certain dapp browsers, or clients?  web3.js(versions of late, anyway) add an id for each receipt log entry.  That `id` is not part of the JSON-RPC spec, but something web3.js does on its own.

That probably shouldn't happen in the webpack compiled contract context, though, as it wouldn't allow a conflict with a global `web3`...

**Comment by mikeshultz** [comment]
Here's an example from web3.js:

    logs:[{
        blockHash: '0x51f070ec5b720941acfaa11b2e54e7dc6b016fdec2d094dc17ccf29a5843ed54',
        transactionHash: '0x91bdbc5eb6ccdac533594a9806bec917865751bc42e6abc38bd889f92f649b58',
        type: 'mined',
        topics: [Array],
        blockNumber: 8111098,
        data: '0x0000000000000000000000000000000000000000000000000000000000000000',
        address: '0x117793CC0b19C01c531638a986923533f2F865E7',
        transactionLogIndex: '0x0',
        logIndex: 95,
        removed: false,
        transactionIndex: 164,
        id: 'log_aea7ddcb'
      }]

And one from raw JSON-RPC:

    "logs": [{
          "blockHash": "0x51f070ec5b720941acfaa11b2e54e7dc6b016fdec2d094dc17ccf29a5843ed54",
          "transactionHash": "0x91bdbc5eb6ccdac533594a9806bec917865751bc42e6abc38bd889f92f649b58",
          "type": "mined",
          "topics": ["0xa38789425dbeee0239e16ff2d2567e31720127fbc6430758c1a4efc6aef29f80"],
          "blockNumber": "0x7bc3fa",
          "data": "0x0000000000000000000000000000000000000000000000000000000000000000",
          "address": "0x117793cc0b19c01c531638a986923533f2f865e7",
          "transactionLogIndex": "0x0",
          "logIndex": "0x5f",
          "removed": false,
          "transactionIndex": "0xa4"
        }]

Only a suspicion though, I'm really not sure why this would ever happen in the wild.

**Comment by franckc** [comment]
Yeah, it's really weird. I added a consistency check to to help confirm this is the issue... let's see !

**Comment by mikeshultz** [comment]
I'm willing to bet @franckc fixed this in PR #2707.  Close?

**Comment by franckc** [comment]
Yeah. Let's close since we haven't seen this happening recently...
